src/Propel/Runtime/ActiveQuery/Criteria.php

```diff
@@ -2867,22 +2843,48 @@ public function doSelect(?ConnectionInterface $con = null)
 
         $params = [];
         $sql = $this->createSelectSql($params);
+        $stmt = null;
         try {
             $stmt = $con->prepare($sql);
             $db->bindValues($stmt, $params, $dbMap);
             $stmt->execute();
         } catch (Exception $e) {
-            if (isset($stmt)) {
-                $stmt = null; // close
-            }
-            Propel::log($e->getMessage(), Propel::LOG_ERR);
-
-            throw new PropelException(sprintf('Unable to execute SELECT statement [%s]', $sql), null, $e);
+            $this->handleStatementException($e, $sql, $con, $stmt);
         }
 
         return $con->getDataFetcher($stmt);
     }
 
+    /**
+     * Logs an exception and adds the complete SQL statement to the exception.
+     *
+     * @param \Exception $e The initial exception.
+     * @param string|null $sql The SQL statement which triggered the exception.
+     * @param \Propel\Runtime\Connection\ConnectionInterface|null $con Used to determine if debug mode is enabled. In debug mode, exception messages
+     *                                  are more verbose, but might unveil sensitive data.
+     * @param \Propel\Runtime\Connection\StatementInterface|\PDOStatement|null $stmt The prepared statement.
+     *
+     * @throws \Propel\Runtime\Exception\PropelException
+     *
+     * @return void
+     */
+    protected function handleStatementException(Exception $e, ?string $sql, ?ConnectionInterface $con = null, $stmt = null): void
+    {
+        $internalMessage = $e->getMessage();
+        Propel::log($internalMessage, Propel::LOG_ERR);
+
+        $useDebugMode = ($con instanceof ConnectionWrapper && $con->useDebug);
```

Comment:
Please extract this into its own method for better readability.

Reply:
Sure. Moved the debug check into its own method.